File: core/src/main/kotlin/org/bitlap/core/sql/table/BitlapSqlQueryMetricTable.kt
package org.bitlap.core.sql.table

import arrow.core.toOption
import org.apache.calcite.DataContext
import org.apache.calcite.linq4j.Enumerable
import org.apache.calcite.linq4j.Linq4j
import org.apache.calcite.plan.RelOptUtil
import org.apache.calcite.rex.RexNode
import org.bitlap.core.data.metadata.Table
import org.bitlap.core.mdm.fetch
import org.bitlap.core.mdm.plan.MetricsMergePlan
import org.bitlap.core.mdm.plan.MetricsPlan
import org.bitlap.core.sql.MDColumnAnalyzer
import org.bitlap.core.sql.PrunePushedFilter
import org.bitlap.core.sql.PruneTimeFilter
import org.bitlap.core.sql.QueryContext

class BitlapSqlQueryMetricTable(
    override val table: Table,
    override val analyzer: MDColumnAnalyzer,
    private val timeFilter: PruneTimeFilter,
    private val pushedFilter: PrunePushedFilter,
) : BitlapSqlQueryTable(table) {

    // filters is empty here, pushed by BitlapFilterTableScanRule
    override fun scan(root: DataContext, filters: MutableList<RexNode>, projects: IntArray?): Enumerable<Array<Any?>> {
        if (projects == null) {
            return Linq4j.emptyEnumerable()
        }
        // TODO: Fix https://github.com/apache/calcite/pull/2729
        val bits = RelOptUtil.InputFinder.bits(filters, null)
        val coverAll = bits.all { projects.contains(it) }
        if (!coverAll) {
            return Linq4j.emptyEnumerable()
        }
        return this.scan(root, projects)
    }

    private fun scan(root: DataContext, projects: IntArray): Enumerable<Array<Any?>> {
        val tbl = this.table
        val rowType = super.getRowType(root.typeFactory)

        val projections = projects.map { rowType.fieldList[it].name!! }
        val metricNames = analyzer.getMetricColNames()
        val dimension = analyzer.getDimensionColNamesWithoutTime()
            .firstOrNull().toOption() // only one or no dimension here
        val dimensionFilter = this.pushedFilter.filter(dimension.orNull())

        val metricParts = metricNames
            .map { analyzer.materializeType(it) }
            .groupBy { it::class.java }

        val rows = fetch {
            queryContext = QueryContext.get()
            table = tbl
            plan = MetricsMergePlan(
                subPlans = metricParts.map { e ->
                    MetricsPlan(timeFilter, e.value, e.key, dimension, dimensionFilter)
                }
            )
        }
        return Linq4j.asEnumerable(rows.toRows(projections))
    }
}

File: bitlap-core/src/main/scala/org/bitlap/core/sql/table/BitlapSqlQueryMetricTable.scala
/*
 * Copyright 2020-2023 IceMimosa, jxnu-liguobin and the Bitlap Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.bitlap.core.sql.table

import java.util.List as JList

import scala.jdk.CollectionConverters.*

import org.bitlap.core.catalog.metadata.Table
import org.bitlap.core.mdm.FetchContext
import org.bitlap.core.mdm.plan.PendingFetchPlan
import org.bitlap.core.sql.PrunePushedFilter
import org.bitlap.core.sql.PruneTimeFilter

import org.apache.calcite.DataContext
import org.apache.calcite.linq4j.Enumerable
import org.apache.calcite.linq4j.Linq4j
import org.apache.calcite.rex.RexNode
import org.apache.calcite.util.ImmutableIntList

class BitlapSqlQueryMetricTable(
  override val table: Table,
  private val timeFilter: PruneTimeFilter,
  private val pushedFilter: PrunePushedFilter)
    extends BitlapSqlQueryTable(table) {

  // filters is empty here, pushed by BitlapFilterTableScanRule
  override def scan(root: DataContext, filters: JList[RexNode], projects: Array[Int]): Enumerable[Array[Any]] = {
    if (projects == null) {
      // return Linq4j.emptyEnumerable()
      val rowType = super.getRowType(root.getTypeFactory)
      this.scan(root, ImmutableIntList.identity(rowType.getFieldCount).toIntArray)
    } else this.scan(root, projects)
  }

  private def scan(root: DataContext, projects: Array[Int]): Enumerable[Array[Any]] = {
    val rowType     = super.getRowType(root.getTypeFactory)
    val projections = projects.map { it => rowType.getFieldList.get(it).getName }.toList

    // add pending fetch plan to context
    val fetchContext = FetchContext(table, PendingFetchPlan(table, analyzer, timeFilter, pushedFilter))

    // find best plan to execute
    val bestPlan = fetchContext.findBestPlan()

    // execute the plan
    val rows = bestPlan.execute(fetchContext)

    Linq4j.asEnumerable(rows.toRows(projections).asJava)
  }
}
